### DISCUSSION TITLE
Water navigation / Measuring distance on a river

### DISCUSSION THREAD

**Discussion by ooprizrakoo** [organicmaps/organicmaps#664] [Ideas]
According to the CIA "The world factbook" worldwide waterways length - 2,293,412 kilometers.
Of course, rivers are not necessarily navigable along the entire length, but for kayakers, fishermen and millions of people living on rivers, it is often important to know the length of the river from its current location to some other location.

It will help you to plan the route taking into account the places for overnight stays, sources of drinking water, and so on.

![rivernav](https://user-images.githubusercontent.com/661813/122677715-2baa4c00-d1ec-11eb-901e-2b05214ae23e.png)

What do you think about?

**Comment by rtsisyk** [comment]
I desperately need this feature too.

**Reply by ramantsitou** [reply]
Very nice idea, I'm looking forward when it will be done.

**Reply by Colonizator1** [reply]
One way to solve this problem is to allow the route engine to be changed to your own, as in the locus. Then we could use the open source brouter and create our own routing profiler including river and others

**Reply by GhostInTheKayak** [reply]
I am a kayaker. This would be very useful.

**Reply by ka2pink** [reply]
Было бы очень полезно и удобно!

**Reply by patepelo** [reply]
> I am a kayaker. This would be very useful.

Have you tried it in osmand or komoot?

**Reply by Quantum-Future** [reply]
Great idea !!
Useful for planning kayak navigation on a river.

**Comment by Rakshith-sakhamuri** [comment]
Hi @ooprizrakoo ,
This is a very nice Idea,I wanted to contribute in this,
I have good experience in C++ ,please guide me for this.
Thank you.

**Reply by AndrewShkrob** [reply]
Check this link, please: [Google Summer of Code 2023](https://github.com/organicmaps/organicmaps/wiki/GSoC-2023-ideas#waterways-navigation-and-routing-rivers-lakes-seas)
OMaps will participate in GSoC and waterways navigation is one of the projects. Probably, you may want to apply.

**Reply by rtsisyk** [reply]
> Hi @ooprizrakoo ,
> 
> This is a very nice Idea,I wanted to contribute in this,
> 
> I have good experience in C++ ,please guide me for this.
> 
> Thank you.

Are you eligible for GSoC? You can start preparing your proposal if we get accepted.

**Reply by MANISH-LAB** [reply]
Hi i have a good experience with c++  and also i worked with similar project where i had to find route and navigation. it was for desert region and i see lot of basic similarities between two, i think i  can contribute to this in a good way, are there any supposedly issues which i have to solve first to get preference for getting selected as contributor?
Thank you.

**Reply by biodranik** [reply]
Choose any issues, we see how contributors are approaching tasks, and will take it into an account for a final decision.

**Comment by pranjalXi** [comment]
Hay @ooprizrakoo ,
I wanted to contribute to this, and I have experience in C++, please guide me with this.

**Reply by biodranik** [reply]
Did you read all our documentation in repo's docs folder? You may start from building the project and fixing some issues. 

Read about OpenStreetMap.org data and learn how waterway's are stored there. 

Then research existing implementations and related issues. 

Then read the code and understand how our routing works.

**Comment by ayushk7102** [comment]
Hello,
I would like to contribute to this issue. I have previously worked on transit routing algorithms, on top of public transit networks built from GTFS feeds. 

I've gone through the relevant documentation, built the project and read about OSM's treatment of waterways (I am familiar with regular OSM-based routing, due to the above project). I'm still reading up on alternative implementations- but had a few questions:

My understanding is that OrganicMaps uses A* as a base algorithm for route planning on road networks-
1. What are some key points (or issues that may be faced) you would recommend thinking about to implement a similar algorithm for waterways?
2. Will this project, as part of a 350-hour project in GSoC '23 also include using Android/iOS implementations of the same? Or just the C++ backend?

Look forward to your response, and contributing to OrganicMaps. Thank you!

**Reply by biodranik** [reply]
1. If you read the code, it becomes clearer. 
2. The working core routing in desktop dev versions is a minimum, if you can add iOS and/or Android UI on top, that would be even better.

**Reply by ayushk7102** [reply]
Great, thank you. Are there any additional points you might suggest to address in my project proposal, if I were to work on this problem statement?

**Reply by biodranik** [reply]
Your proposal should reflect your understanding of our code.

**Comment by lalithreddy1** [comment]
Hi @ooprizrakoo  This is a very good idea , I have some prior experience working with maps where i implemented the same thing but where we worked on roads. We  heard a problem that Travelling salesman have an issue when they want to visit many locations in a trip and they want to check the shortest route along with the resting spots and good areas of food , so that they can plan a trip in a effective way. We did this is Python Programming language and I am pretty sure that we can do the same in C++ .  I want to contribute in to this open source project. Please share me some of the sources of the project so that i can  easily understand the WIP .

**Comment by M-INDN-SEDTA** [comment]
Hi
This is interesting and helpful for people.
I know C++ and DSA, can complete this project within the time.
Allow me to do this project as 'SEDTA' via GSOC proposal.

**Comment by junyizha** [comment]
Hi,
I am David, a senior majoring in computer science at Carnegie Mellon University. I've used C++ extensively for implementing connection pools in a trading system and implementing a complete database management system. I would like to confirm that this issue is still open before I start working on the issue. I've played around with Organic Maps for a while and I'll start looking at the codebase to brainstorm solutions.

**Reply by alexey-naumenko** [reply]
Hello, this task has not been started yet.
We will be glad to see you in our chat - https://t.me/organicmaps or if you write for details directly to the developers @rtsisyk, e.g.
Thanks.

**Reply by patepelo** [reply]
> Hi,
> I am David, a senior majoring in computer science at Carnegie Mellon University. I've used C++ extensively for implementing connection pools in a trading system and implementing a complete database management system. I would like to confirm that this issue is still open before I start working on the issue. I've played around with Organic Maps for a while and I'll start looking at the codebase to brainstorm solutions.

Hey David, you're very welcome to contribute. Almost certainly nobody is working on this. You can start a draft and ask the community if needed ✌️

**Comment by mrtootoo** [comment]
Great idea.  In addition, I would also like to see other features such as Navaids, water depths, and other boating POIs.  All this info is available for free.  I think there is even an Open Sea Map.

**Reply by patepelo** [reply]
> Great idea.  In addition, I would also like to see other features such as Navaids, water depths, and other boating POIs.  All this info is available for free.  I think there is even an Open Sea Map.

This is a good idea. Please open its own Issue if you don't find one already. That way we can start discussing elements for boating.

**Reply by msturtz1** [reply]
This please -- I came to request exactly this.  The commercial nautical map tools are stupid expensive for what you actually get, unless you're on a yacht.  OpenSeaMap is somewhat limited to Europe and Baltics last I checked, but with an awesome client perhaps that would change.  I can contribute a few bucks but, but I have no other ability to support.

Sounds like a new request was requested?  Has that happened?

**Comment by biodranik** [comment]
Note that the existing "Ruler" routing option allows building a rough linear route with many stop points on the river.

**Reply by msturtz1** [reply]
Sure, to measure distance -- but nautical route planning/finding and boating would be the ultimate goal here

**Comment by mrtootoo** [comment]
Could GEBCO provide the bathymetric data needed?

**Comment by wildsurfer** [comment]
Hi all — been using Organic Maps since the maps.me days, so this thread caught my eye.

I run an open source project called [auto-sea-way](https://github.com/auto-sea-way/asw) that's at least partly relevant here. It's a maritime auto-router written in Rust. The build step takes OSM land polygons and produces a global water-surface graph using H3 hexagons, with the resolution adapting from res-3 in the open ocean down to res-13 in tight spots like Suez and Panama. The output is one binary file (bitcode + zstd) and there's a small HTTP server on top that does A* and answers `/route`. MIT/Apache-2.0.

One thing worth flagging though: the original post here is really about rivers, and rivers are a different problem than what asw solves. In OSM, rivers are polylines tagged `waterway=river`, not polygons, so the "water = anything that isn't land" trick asw uses doesn't carry over. River routing is closer to how OM already does roads — you walk the waterway network and stitch connected segments. Where asw is genuinely useful is the open-water side that came up later in the thread (boating, OpenSeaMap, etc.).

That said, a fair amount of the surrounding work transfers: H3 indexing, nearest-node lookup, A* with reusable buffers, the OSM ingest, the build pipeline on a Hetzner box. If someone on the OM side ends up actually picking this up, I'd be glad to compare notes — what worked, what hurt, where the cliffs are. Easiest to reach me on the repo.

Cheers, and thanks for keeping maps.me alive all these years.
